### PR TITLE
Hide ECX_KEY again

### LINKED ISSUE #12880: The internal ECX_KEY structure seems to be exposed publicly
The ECX_KEY structure is a purely internal type. We have never exposed a low level API for ECX keys. However we seem to have exposed this type in the public API (and documented them) recently via the function calls `EVP_PKEY_get0_X25519()`, `EVP_PKEY_get0_X448()`, `EVP_PKEY_get0_ED25519()`, `EVP_PKEY_get0_ED448()` as well as the equivalent `set0` variants.

This doesn't seem correct - they should be removed.

### PR DESCRIPTION
**Pull request by levitte**
ECX_KEY was not meant for public consumption, it was only to be
accessed indirectly via EVP routines.  However, we still need internal
access for our decoders.

This partially reverts 7c664b1f1b5f60bf896f5fdea5c08c401c541dfe

Fixes #12880

### CODE REVIEW

**Comment by t-j-h** [comment]
How does one work with these types of keys at the EVP_PKEY level?
I'm assuming this is a violate-current-patterns on the grounds of not digging the hole deeper?

**Comment by levitte** [comment]
> I'm assuming this is a violate-current-patterns on the grounds of not digging the hole deeper?

At some point, it was decided that we would not publicly expose new low-level keys, or functions.  They would only be publicly accessible via EVP_PKEY.  ECX_KEY is such a type.
With that, the pattern with assigning low-level keys to and extracting them from EVP_PKEY isn't a current pattern as much as an old pattern.

**Review by slontis — COMMENTED** [review]
Was in the process of doing this.. I even assigned it to myself. Not to worry.

File: include/crypto/ecx.h [hunk 1]
```diff
@@ -109,10 +109,19 @@ int X448(uint8_t out_shared_key[56], const uint8_t private_key[56],
 void X448_public_from_private(uint8_t out_public_value[56],
                               const uint8_t private_key[56]);
 
+
 /* Backend support */
 int ecx_public_from_private(ECX_KEY *key);
 int ecx_key_fromdata(ECX_KEY *ecx, const OSSL_PARAM params[],
                      int include_private);
 
+ECX_KEY *evp_pkey_get0_X25519(const EVP_PKEY *pkey);
```
Comment (slontis) [inline include/crypto/ecx.h:118]: These seems like overkill. This is all we actually need..
```
ECX_KEY *evp_pkey_get1_ECX_KEY(EVP_PKEY *pkey, int type)
```

The encoder can do this
```
#define IMPLEMENT_ECX_NEWCTX(KEYTYPEstr, KEYTYPE, keytype, free)    \
static ECX_KEY *evp_pkey_get1_##keytype(EVP_PKEY *pkey)                        \
{                                                                              \
    return evp_pkey_get1_ECX_KEY(pkey, EVP_PKEY_##KEYTYPE);                    \
}                                                                              \
IMPLEMENT_NEWCTX(KEYTYPEstr, KEYTYPE, keytype, evp_pkey_get1_##keytype, free)
```

Comment (slontis) [inline include/crypto/ecx.h:118]: This either needs to go in crypto/evp.h or it needs to be renamed to
ecx_key_

File: util/libcrypto.num [hunk 2]
```diff
@@ -5163,18 +5163,6 @@ ERR_load_OSSL_DECODER_strings           ?	3_0_0	EXIST::FUNCTION:
...
 OSSL_DECODER_CTX_new_by_EVP_PKEY        ?	3_0_0	EXIST::FUNCTION:
-EVP_PKEY_set1_X25519                    ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_get0_X25519                    ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_get1_X25519                    ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_set1_X448                      ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_get0_X448                      ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_get1_X448                      ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_set1_ED25519                   ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_get0_ED25519                   ?	3_0_0	EXIST::FUNCTION:EC
-EVP_PKEY_get1_ED25519                   ?	3_0_0	EXIST::FUNCTION:EC
```
Comment (slontis) [inline util/libcrypto.num:5174]: Will deleting these cause issues?

File: include/crypto/ecx.h (hunk 1 above)
Comment (levitte) [reply include/crypto/ecx.h:118]: I disagree...  they return ECX_KEY pointers, so they can be claimed to fit here just as well.
Furthermore, `crypto/evp.h` contains a little too much stuff that can lead to undesired shortcuts, and the decoders already include `crypto/ecx.h`.  Also, if you look in files like `include/openssl/ec.h`, you'll find EVP functions / macros there.
Also, considering this is internal anyway, I don't see a need to be overly fussier...

File: util/libcrypto.num (hunk 2 above)
Comment (levitte) [reply util/libcrypto.num:5174]: Well, er, no.  The whole idea with this PR is exactly this, to unpublish what was published by mistake

Comment (levitte) [reply util/libcrypto.num:5174]: Or, it *shouldn't* be an issue.  If there is, I need to tinker this PR a little more

Comment (levitte) [reply util/libcrypto.num:5174]: I just checked all the builds, and they don't indicate any problem.  So...

File: include/crypto/ecx.h (hunk 1 above)
Comment (levitte) [reply include/crypto/ecx.h:118]: Yeahok...  I was again thinking in "feature complete" terms, which is a bit silly for internal functions, I have to admit.

Comment (levitte) [reply include/crypto/ecx.h]: Fixed

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
Merged to master.

**Pull request closed without merging**